src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/AlleleSubsettingUtils.java

```diff
@@ -0,0 +1,421 @@
+package org.broadinstitute.hellbender.tools.walkers.genotyper;
+
+import htsjdk.variant.variantcontext.*;
+import org.broadinstitute.hellbender.utils.MathUtils;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.variant.GATKVCFConstants;
+import org.broadinstitute.hellbender.utils.variant.GATKVariantContextUtils;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Created by davidben on 6/8/16.
+ */
+public class AlleleSubsettingUtils {
+    private AlleleSubsettingUtils() {}  //prevent instantiation
+    private static final int PL_INDEX_OF_HOM_REF = 0;
+    private static final int MAX_LENGTH_FOR_PL_LOGGING = 100; // if PL vectors longer than this # of elements, don't log them
+
+    /**
+     * From a given variant context, extract a given subset of alleles, and update genotype context accordingly,
+     * including updating the PL's, and assign genotypes accordingly
+     * @param vc                                variant context with alleles and genotype likelihoods
+     * @param defaultPloidy                     ploidy to use when {@code vc} is missing information for a sample
+     * @param allelesToKeep                     subset of alleles in {@code vc} to keep
+     * @param method                            whether to assign hard genotypes or leave as no-call
+     * @return                                  GenotypesContext with new PLs
+     */
+    public static GenotypesContext subsetAlleles(final VariantContext vc, final int defaultPloidy, final boolean isDiploid,
+                                                              final List<Allele> allelesToKeep, final GenotypeAssignmentMethod method) {
+        return isDiploid ? subsetDiploidAlleles(vc, allelesToKeep, method) :
+                subsetGeneralPloidyAlleles(vc, defaultPloidy, allelesToKeep, method);
+    }
+
+    /**
+     * Restrict a variant context to its "best" alleles.
+     * @param vc                    variant context with alleles and genotype likelihoods
+     * @param defaultPloidy         ploidy to assume in case that {@code vc} does not contain that information
+     * @param numAltAllelesToKeep
+     * @param isDiploid
+     * @return
+     */
```

Comment:
mmm... this is a bit convoluted.. for example what is the actual difference between ```subsetAllele``` and ```reduceNumberOfAlleles```. Yes I know, it is the return type how you specify the alleles to keep .... yet It looks a bit wrong... On the one hand I would expect that these methods would share the name but differ in the signature yet they have different return types which complicate things... I think part of the problem is that Genotypes do not hold a reference to their enclosing VC thus the allele list. 

I need to think this through.